### PR TITLE
[SPARK-28478][SQL] Remove redundant null checks

### PR DESCRIPTION
**Pull request by davidvrba**
### What changes were proposed in this pull request?
The purpose of this pr is to remove explicit null checks if they are not needed in order to simplify the generated code. Here is one example: 

Expressions of this type
```
CASE WHEN isnull(title#5) THEN title#5 ELSE substring(title#5, 0, 3) END
```
are simplified to 
```
substring(title#5, 0, 3)
```
if the considered expression is null-intolerant.

### Why are the changes needed?
It simplifies expressions in the query plan which leads to potential optimization due to simplified codegen.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
New tests are added.

### CODE REVIEW

**Comment by davidvrba** [comment]
@cloud-fan kindly asking for review. Thanks for the help.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -757,3 +757,45 @@ object CombineConcats extends Rule[LogicalPlan] {
...
+      ifNullExpr: Expression,
+      ifNotNullExpr: Expression,
+      checkedExpr: Expression): Boolean = {
+    (ifNullExpr == checkedExpr || ifNullExpr == Literal.create(null, checkedExpr.dataType)) &&
+      ifNotNullExpr.isInstanceOf[NullIntolerant] &&
+      ifNotNullExpr.children.contains(checkedExpr)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+    case i @ If(predicate, trueValue, falseValue) => predicate match {
```
[the remark this answers is not in the document]
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: The `NullPropagation` rule simplifies expressions to literals. But i feel that my pr is covering slightly different case. Here the expression that is being null-checked is in general not `Literal` and can not be converted to `Literal` (in general). 
However I can also see that the logic of my rule can be moved to `SimplifyConditionals`, so I can move it there if this is the preferred way.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: yes please.

[the remark this answers is not in the document]
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: ok, it is moved

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -434,6 +434,22 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
...
+   */
+  private def isRedundantNullCheck(
+      ifNullExpr: Expression,
+      ifNotNullExpr: Expression,
+      checkedExpr: Expression): Boolean = ifNotNullExpr match {
+    case e: NullIntolerant if (
+      (ifNullExpr == checkedExpr || ifNullExpr == Literal.create(null, checkedExpr.dataType))
+      && e.children.contains(checkedExpr)) => true
+    case _ => false
+  }
```
[the remark this answers is not in the document]
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Makes sense.

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Can you generalize the last condition more? e.g., how about the case, `substring(other_func(title#5), 0, 3)` in the example you described?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -442,6 +459,13 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(cond, trueValue, falseValue)
         if cond.deterministic && trueValue.semanticEquals(falseValue) => trueValue
 
+      case i @ If(cond, trueValue, falseValue) => cond match {
+        // If the null-check is redundant, remove it
+        case IsNull(child) if isRedundantNullCheck(trueValue, falseValue, child) => falseValue
+        case IsNotNull(child) if isRedundantNullCheck(falseValue, trueValue, child) => trueValue
+        case _ => i
+      }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: How about this format?;
```
      // If the null-check is redundant, remove it
      case If(IsNull(child), trueValue, falseValue)
        if isRedundantNullCheck(trueValue, falseValue, child) => falseValue
      case If(IsNotNull(child), trueValue, falseValue)
        if isRedundantNullCheck(falseValue, trueValue, child) => trueValue
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 4]
```diff
@@ -483,6 +507,19 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
...
+        branches(0)._1 match {
+          case IsNotNull(child) if isRedundantNullCheck(
+            elseValue.getOrElse(Literal.create(null, child.dataType)),
+            branches(0)._2, child) => branches(0)._2
+          case IsNull(child) if isRedundantNullCheck(
+            branches(0)._2,
+            elseValue.getOrElse(Literal.create(null, child.dataType)),
+            child) => elseValue.getOrElse(Literal.create(null, child.dataType))
+          case _ => e
+        }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: How about this?
```
      // remove redundant null checks for CaseWhen with one branch
      case CaseWhen(Seq((IsNotNull(child), trueValue)), Some(falseValue))
        if isRedundantNullCheck(falseValue, trueValue, child) => trueValue
      case CaseWhen(Seq((IsNull(child), trueValue)), Some(falseValue))
        if isRedundantNullCheck(trueValue, falseValue, child) => falseValue
      case CaseWhen(Seq((IsNotNull(child), trueValue)), None)
        if isRedundantNullCheck(Literal.create(null, child.dataType), trueValue, child) => trueValue
      case e @ CaseWhen(Seq((IsNull(child), trueValue)), None) =>
        val nullValue = Literal.create(null, child.dataType)
        if (isRedundantNullCheck(trueValue, nullValue, child)) {
          nullValue
        } else {
          e
        }
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala [hunk 5]
```diff
@@ -35,18 +35,24 @@ class SimplifyConditionalSuite extends PlanTest with PredicateHelper {
   }
 
   protected def assertEquivalent(e1: Expression, e2: Expression): Unit = {
-    val correctAnswer = Project(Alias(e2, "out")() :: Nil, OneRowRelation()).analyze
-    val actual = Optimize.execute(Project(Alias(e1, "out")() :: Nil, OneRowRelation()).analyze)
+    val correctAnswer = Project(Alias(e2, "out")() :: Nil, LocalRelation('a.int)).analyze
+    val actual = Optimize.execute(
+      Project(Alias(e1, "out")() :: Nil, LocalRelation('a.int)).analyze)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: nit: you don't need to break this line.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala [hunk 6]
```diff
@@ -18,15 +18,15 @@
...
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.Literal.{FalseLiteral, TrueLiteral}
 import org.apache.spark.sql.catalyst.plans.PlanTest
-import org.apache.spark.sql.catalyst.plans.logical._
-import org.apache.spark.sql.catalyst.rules._
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
 import org.apache.spark.sql.types.{IntegerType, NullType}
 
-
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: nit: You need to avoid unnecessary changes like this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 7]
```diff
@@ -483,6 +507,19 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case e @ CaseWhen(branches, elseValue) if branches.length == 1 =>
+        // remove redundant null checks for CaseWhen with one branch
+        branches(0)._1 match {
+          case IsNotNull(child) if isRedundantNullCheck(
+            elseValue.getOrElse(Literal.create(null, child.dataType)),
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: `child.dataType` -> `e.dataType`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala [hunk 8]
```diff
@@ -35,18 +35,24 @@ class SimplifyConditionalSuite extends PlanTest with PredicateHelper {
...
   }
 
   private val trueBranch = (TrueLiteral, Literal(5))
   private val normalBranch = (NonFoldableLiteral(true), Literal(10))
   private val unreachableBranch = (FalseLiteral, Literal(20))
-  private val nullBranch = (Literal.create(null, NullType), Literal(30))
-
-  val isNotNullCond = IsNotNull(UnresolvedAttribute(Seq("a")))
-  val isNullCond = IsNull(UnresolvedAttribute("b"))
+  private val nullValue = Literal.create(null, IntegerType)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: Why did you change from `NullType` to `IntegerType` here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala [hunk 9]
```diff
@@ -128,7 +192,7 @@ class SimplifyConditionalSuite extends PlanTest with PredicateHelper {
     // When the conditions in `CaseWhen` are all deterministic, `CaseWhen` can be removed.
     assertEquivalent(
       CaseWhen((isNotNullCond, Subtract(Literal(3), Literal(2))) ::
-        (isNullCond, Literal(1)) ::
+        (isNullCondB, Literal(1)) ::
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: You don't need to change the existing tests where possible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: yes, that should be possible.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala (hunk 6 above)
Comment (davidvrba) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: Ok

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala (hunk 5 above)
Comment (davidvrba) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: I see

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 4 above)
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Alright

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala (hunk 8 above)
Comment (davidvrba) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: I need the same dataType as i have for the `a` attribute. But i can just add another nullValue to the test and keep the previous with the original dataType.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala (hunk 9 above)
Comment (davidvrba) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: Ok, i will try to avoid that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 3 above)
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Ok

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala (hunk 8 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalSuite.scala]: Yea, I think its better to avoid the behaviour changes in the existing tests.

**Comment by davidvrba** [comment]
@cloud-fan @maropu  do you have any more suggestions / comments / recommendations to this pr? In the last commit i added a bit of generalization to include cases where the null-checked column is not necessarily a _direct_ child of the `ifNotNullExpr`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 10]
```diff
@@ -434,6 +434,27 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
...
+   * @param ifNullExpr expression that takes place if checkedExpr is null
+   * @param ifNotNullExpr expression that takes place if checkedExpr is not null
+   * @param checkedExpr expression that is checked for null value
+   */
+  private def isRedundantNullCheck(
+      ifNullExpr: Expression,
+      ifNotNullExpr: Expression,
+      checkedExpr: Expression): Boolean = {
+    val isNullIntolerant = ifNotNullExpr.find { x =>
+      !x.isInstanceOf[NullIntolerant] && x.find(e => e.semanticEquals(checkedExpr)).nonEmpty
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: The same logic? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala#L105-L109

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Why did you add the inner pattern-matching (`cond match {` )? I think its better to avoid unnecessary pattern matching (In the current fix, all the cases for `If` exprs can be matched in the line 466).

Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: I see. You are right, i do not need the inner pattern match, i will fix that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 10 above)
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Actually i think we need slightly different logic. Consider these two examples where `x` will be the null-checked column: 

1. `substring(x, coalesce(a, b), c)`

2. `substring(coalesce(x, d), a, c)`

For 1. we need to be null-intolerant (even though `coalesce` is null-tolerant), so if `x` is null, we replace the `substring` with null value no matter what are the other children. For 2. we need to be null-tolerant and we will not replace the `substring` by null value. So we need to check the expression with respect to the position of `x` (the column that is being null-checked). Does it make sense?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Probably, you meant `FiterExec.isNullIntolerant(ifNotNullExpr) || additional checks for the case having null-tolerant exprs inside ifNotNullExpr`? (`FiterExec.isNullIntolerant` is private though...)

Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Yeah, `FilterExec.isNullIntolerant(ifNotNullExpr)` is a stronger condition than we need so in case there is null-tolerant expr inside we need to check if the null-checked column is in its subtree. Using the logic from `FilterExec.isNullIntolerant` the function could look like this:
```
def isNullIntolerant(expr: Expression): Boolean = expr match {
  case e: NullIntolerant => e.children.forall(isNullIntolerant)
  case e if e.find(x => x.semanticEquals(checkedExpr)).isEmpty => true
  case _ => false
}
```

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Ah, I see. For better code readability, could you split the condition into the two parts as I suggested above? Also, I think its better to leave some comments about why we need more checks there.

Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: I agree that the committed code is not very intuitive so i can think of this way which seems to be more readable (added also some comments):

```
private def isRedundantNullCheck(
    ifNullExpr: Expression,
    ifNotNullExpr: Expression,
    checkedExpr: Expression): Boolean = {
  
  // checks if expr is null-intolerant with respect to checkedExpr
  def isNullIntolerant(expr: Expression): Boolean = expr match {
    case e: NullIntolerant => e.children.forall(isNullIntolerant)
    // if some child is null-tolerant but the checkedEpxr is not in its subtree
    // we can still consider the whole expr as null-intolerant 
    // with respect to checkedExpr
    case e if e.find(x => x.semanticEquals(checkedExpr)).isEmpty => true
    case _ => false
  }
  
  isNullIntolerant(ifNotNullExpr) && {
    (ifNullExpr.semanticEquals(checkedExpr) ||
      (ifNullExpr.foldable && ifNullExpr.eval() == null)) &&
      // we still need to make sure that checkedExpr is inside ifNotNullExpr
      ifNotNullExpr.find(x => x.semanticEquals(checkedExpr)).nonEmpty
  }
}
```
But not sure if this is what you had in mind when suggesting to split the condition. Can you think of a better way how to compose this?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Hmmm, that still looks complicated.. If we cannot avoid the complexity for the stronger condition, as another option, I think we can cover the simple case (`FiterExec.isNullIntolerant(ifNotNullExpr)`) only in this pr. If necessary, we might be able to optimize the condition in future work. I think keeping the code simple is more important. WDYT?

Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Well, the think is that if we use the simple version with `FilterExec.isNullIntolerant(ifNutNullExpr)` we will loose (because of the recursive check) all expressions that contain literals (because literals are null-tolerant), so for example expressions like this `substring(title#5, 0, 3)` will not be included in the optimization (which the jira was targeted for in the first place). So I suggest one of these 2 options:

1. Use the complex version of the code and thus include more expressions in the optimization
2. Have the code more simple and use the original version before the generalization step, i.e. 
```
private def isRedundantNullCheck(
    ifNullExpr: Expression,
    ifNotNullExpr: Expression,
    checkedExpr: Expression): Boolean = {
  ifNotNullExpr.isInstanceOf[NullIntolerant] && {
    (ifNullExpr == checkedExpr || ifNullExpr == Literal.create(null, checkedExpr.dataType)) &&
      ifNotNullExpr.children.contains(checkedExpr)
  }
}
```
where `checkedExpr` must be direct child and thus we don't have to check the whole subtree for null-intolerance (so expressions that have Literals in the subtree are still included).
I am fine with either of these 2 options. What do you think?
